mm2src/common/jsonrpc_client.rs

```diff
@@ -334,133 +350,94 @@ pub trait JsonRpcMultiClient: JsonRpcClient {
         request: JsonRpcRequest,
     ) -> RpcRes<T> {
         let client_info = self.client_info();
+        let request = JsonRpcRequestEnum::Single(request);
         Box::new(
-            self.transport_exact(to_addr.to_owned(), JsonRpcRequestEnum::Single(request.clone()))
-                .then(move |result| process_transport_single_result(result, client_info, request)),
+            self.transport_exact(to_addr.to_owned(), request.clone())
+                .then(move |result| process_transport_single_result(result))
+                .map_err(move |e| JsonRpcError::new(&client_info, request, e)),
         )
     }
 }
 
 /// Checks if the given `result` is success and contains `JsonRpcResponse`.
 /// Tries to decode the batch response into `T`.
 fn process_transport_single_result<T: DeserializeOwned + Send + 'static>(
-    result: Result<(JsonRpcRemoteAddr, JsonRpcResponseEnum), String>,
-    client_info: String,
-    request: JsonRpcRequest,
-) -> Result<T, JsonRpcError> {
-    let request = JsonRpcRequestEnum::Single(request);
-
+    result: Result<(JsonRpcRemoteAddr, JsonRpcResponseEnum), JsonRpcErrorType>,
+) -> Result<T, JsonRpcErrorType> {
     match result {
-        Ok((remote_addr, JsonRpcResponseEnum::Single(single))) => {
-            process_single_response(client_info, remote_addr, request, single)
-        },
-        Ok((remote_addr, JsonRpcResponseEnum::Batch(batch))) => {
-            let error = ERRL!("Expeced single response, found batch response: {:?}", batch);
-            Err(JsonRpcError {
-                client_info,
-                request,
-                error: JsonRpcErrorType::Parse(remote_addr, error),
-            })
-        },
-        Err(e) => Err(JsonRpcError {
-            client_info,
-            request,
-            error: JsonRpcErrorType::Transport(e),
-        }),
+        Ok((remote_addr, JsonRpcResponseEnum::Single(single))) => process_single_response(remote_addr, single),
+        Ok((remote_addr, JsonRpcResponseEnum::Batch(batch))) => Err(JsonRpcErrorType::Parse(
+            remote_addr,
+            ERRL!("Expeced single response, found batch response: {:?}", batch),
```

Comment:
typo, should be `ERRL!("Expected single response, found batch response: {:?}", batch),`